stan/math/prim/mat/fun/append_array.hpp

```diff
@@ -1,77 +1,76 @@
-      z.insert(z.end(), y.begin(), y.end());
-      return z;
-    }
-
-    /**
-     * Return the concatenation of two vectors containing Matrices
-     *   of the same type in the order of the arguments.
+     * The return type is computed with append_return_type
      *
-     * @tparam T1 Type of element in Matrices in first vector
-     * @tparam T2 Type of element in Matrices in second vector
-     * @tparam R Row specification of matrices
-     * @tparam C Column specification of matrices
+     * @tparam T1 Scalar type of first vector
+     * @tparam T2 Scalar type of second vector
      * @param x First vector
      * @param y Second vector
      * @return A vector of x and y concatenated together (in that order)
      */
-    template <typename T1, typename T2, int R, int C>
-    inline std::vector<Eigen::Matrix<typename return_type<T1, T2>::type, R, C> >
-    append_array(const std::vector<Eigen::Matrix<T1, R, C> >& x,
-                 const std::vector<Eigen::Matrix<T2, R, C> >& y) {
-      if (!x.empty() && !y.empty()) {
-        check_matching_dims("append_array",
-                            "x[1]",
-                            x.front(),
-                            "y[1]",
-                            y.front());
+    template <typename T1, typename T2 >
+    inline typename append_return_type<std::vector<T1>, std::vector<T2> >::type
+    append_array(const std::vector<T1>& x, const std::vector<T2>& y) {
+      typename append_return_type<std::vector<T1>, std::vector<T2> >::type z;
+      std::vector<int> zdims;
+      if (x.empty()) {
+        zdims = dims(y);
+        zdims[0] += x.size();
+      } else {
+        zdims = dims(x);
+        zdims[0] += y.size();
       }
-
-      std::vector<Eigen::Matrix<typename return_type<T1, T2>::type, R, C> > z;
-      z.reserve(x.size() + y.size());
-      for (size_t i = 0; i < x.size(); i++)
-        z.push_back(x[i].template cast<typename return_type<T1, T2>::type>());
-      for (size_t i = 0; i < y.size(); i++)
-        z.push_back(y[i].template cast<typename return_type<T1, T2>::type>());
+      resize(z, zdims);
+      for (size_t i = 0; i < x.size(); ++i)
+        assign(z[i], x[i]);
+      for (size_t i = 0; i < y.size(); ++i)
+        assign(z[i + x.size()], y[i]);
       return z;
     }
 
     /**
      * Return the concatenation of two specified vectors in the order of
      *   the arguments.
      *
-     * @tparam T1 Scalar type of first vector
-     * @tparam T2 Scalar type of second vector
+     * @tparam T1 Type of vectors
      * @param x First vector
      * @param y Second vector
      * @return A vector of x and y concatenated together (in that order)
      */
-    template <typename T1, typename T2>
-    inline typename std::vector<typename return_type<T1, T2>::type>
-    append_array(const std::vector<T1>& x, const std::vector<T2>& y) {
-      std::vector<typename return_type<T1, T2>::type> z;
+    template <typename T1>
```

Comment:
I agree with another comment somewhere that a lot of this should just go in the doc itself.